src/Streamly/Internal/Data/Stream/StreamD.hs

```diff
@@ -4034,3 +4041,65 @@ lastN n = Fold step initial done
         foldFunc i
             | i < n = RB.unsafeFoldRingM
             | otherwise = RB.unsafeFoldRingFullM
+
+------------------------------------------------------------------------------
+-- Time related
+------------------------------------------------------------------------------
+
+{-# INLINE secondsToTimeSpec #-}
+secondsToTimeSpec :: Double -> TimeSpec
+secondsToTimeSpec d = TimeSpec (a :: Int64) (x :: Int64)  
+    where
+        tenPow9 = 10^(9 :: Int)
+        (a, b) = properFraction d
+        x = round $ b * tenPow9
+
+-- XXX This does not fuse. Removing the if function results in fusion though.
+-- XXX This function will fuse with the use of fusion-plugin.
+-- XXX This function also fuses if compiled with -funfolding-use-threshold=500
+{-# INLINE_NORMAL takeByTime #-}
+takeByTime :: MonadIO m => Double -> Stream m a -> Stream m a
```

Comment:
A `Double` seconds requires `TimeSpec` to hold it. Arithmetic on `TimeSpec` may be slow, this could be expensive since we have to do it in each iteration of the loop, for faster operations we may want to restrict the time argument to `TimeUnit64` which can do faster arithmetic.

However, currently all the cost is masked by the `getTime` call, so using efficient time arithmetic may not have any significant impact. But if we use a less costly time acquisition function (coarse granularity) then it may start to matter.

Reply:
I think we make the signature `MonadIO m, TimeUnit64 t => ...` instead of using `Double`.
This is more clear and simple than converting time from `Double` (seconds). We can export it once we decide on `Units` for time.